File: androlua/src/main/java/cn/vove7/androlua/LuaHelper.java
package cn.vove7.androlua;

import android.content.Context;
import android.util.Log;

import cn.vove7.androlua.androbridge.BridgeManager;
import cn.vove7.androlua.luabridge.LuaThread;
import cn.vove7.androlua.luabridge.LuaUtil;
import cn.vove7.androlua.luautils.LuaGcable;
import com.luajava.JavaFunction;
import com.luajava.LuaException;
import com.luajava.LuaState;
import com.luajava.LuaStateFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Pattern;

import cn.vove7.androlua.luautils.LuaDexLoader;
import cn.vove7.androlua.luautils.LuaManagerI;
import cn.vove7.androlua.luautils.LuaPrinter;
import dalvik.system.DexClassLoader;

import static cn.vove7.androlua.luabridge.LuaUtil.errorReason;
import static com.luajava.LuaState.LUA_GCSTOP;

/**
 * LuaHelper
 * <p>
 * Created by Vove on 2018/7/30
 */
public class LuaHelper implements LuaManagerI {
    private Context context;
    public LuaState L;
    private String libDir;
    private String jniLibsPath;
    private String luaDir;
    private ArrayList<LuaGcable> gcList;
    private String luaRequireSearchPath;
    private LuaDexLoader mLuaDexLoader;
    private BridgeManager bridgeManager;

    public BridgeManager getBridgeManager() {
        return bridgeManager;
    }

    public void setBridgeManager(BridgeManager bridgeManager) {
        this.bridgeManager = bridgeManager;
    }

    public LuaHelper(Context context) {
        this.context = context;
        initPath();
        init();
    }

    private void init() {
        gcList = new ArrayList<>();
        mLuaDexLoader = new LuaDexLoader(context);
        try {
            mLuaDexLoader.loadLibs();
        } catch (LuaException e) {
            e.printStackTrace();
        }
        initLua();
    }

    private void initPath() {
        luaDir = context.getFilesDir().getAbsolutePath();
        Log.d("Vove :", "initLua luaDir ----> " + luaDir);
        libDir = context.getDir("lib", Context.MODE_PRIVATE).getAbsolutePath();
        jniLibsPath = context.getApplicationInfo().nativeLibraryDir + "/lib?.so" + ";" + libDir + "/lib?.so;";

        String assetPath = "file://android_assset";
        String assetInclude = assetPath + "/?.lua;" + assetPath + "/?/init.lua;";
        luaRequireSearchPath = (luaDir + "/?.lua;" + luaDir + "/lua/?.lua;" + luaDir + "/?/init.lua;");
        luaRequireSearchPath += jniLibsPath + assetInclude;
    }

    @Override
    public void log(String log) {
        Log.d("Vove :", "lua log  ----> " + log);
    }

    private void initLua() {
        L = LuaStateFactory.newLuaState();
        L.openLibs();
        L.pushJavaObject(this);
        L.setGlobal("luaman");
        L.getGlobal("luaman");
        L.pushContext(context);
        //L.setGlobal("app");

        L.getGlobal("luajava");

        //L.pushString(luaDir);
        //L.setField(-2, "luaextdir");
        L.pushString(luaDir);
        L.setField(-2, "luaDir");
        //L.pushString(luaDir);
        //L.setField(-2, "luapath");
        L.pop(1);

        L.getGlobal("package");
        L.pushString(luaRequireSearchPath);
        L.setField(-2, "path");
        L.pushString(jniLibsPath);
        L.setField(-2, "cpath");
        L.pop(1);

        new LuaPrinter(L, new LuaPrinter.OnPrint() {
            @Override
            public void onPrint(int l, String output) {
                notifyOutput(l, output);
            }
        });

        JavaFunction set = new JavaFunction(L) {
            @Override
            public int execute() throws LuaException {
                LuaThread thread = (LuaThread) L.toJavaObject(2);

                thread.set(L.toString(3), L.toJavaObject(4));
                return 0;
            }
        };

        JavaFunction call = new JavaFunction(L) {
            @Override
            public int execute() throws LuaException {
                LuaThread thread = (LuaThread) L.toJavaObject(2);

                int top = L.getTop();
                if (top > 3) {
                    Object[] args = new Object[top - 3];
                    for (int i = 4; i <= top; i++) {
                        args[i - 4] = L.toJavaObject(i);
                    }
                    thread.call(L.toString(3), args);
                } else if (top == 3) {
                    thread.call(L.toString(3));
                }
                return 0;
            }
        };
        try {
            set.register("set");
            call.register("call");
        } catch (LuaException e) {
            e.printStackTrace();
        }

    }

    @Override
    public void handleMessage(int l, String msg) {
        notifyOutput(l,msg);
    }

    //private void initLua1() {
    //
    //    try {
    //        L = LuaStateFactory.newLuaState();
    //        L.openLibs();
    //
    //        L.pushJavaObject(context);
    //        L.setGlobal("this");
    //
    //        JavaFunction assetLoader = new LuaAssetLoader(context, L);
    //
    //        L.getGlobal("package");            // package
    //        L.getField(-1, "loaders");         // package loaders
    //        int nLoaders = L.objLen(-1);       // package loaders
    //
    //        L.pushJavaFunction(assetLoader);   // package loaders loader
    //        L.rawSetI(-2, nLoaders + 1);       // package loaders
    //        L.pop(1);                          // package
    //
    //        L.getField(-1, "path");            // package path
    //        String customPath = context.getFilesDir() + "/?.lua";
    //        L.pushString(";" + customPath);    // package path custom
    //        L.concat(2);                       // package pathCustom
    //        L.setField(-2, "path");            // package
    //        L.pop(1);
    //
    //    } catch (Exception e) {
    //        e.printStackTrace();
    //    }
    //}

    void safeEvalLua(String src) {
        try {
            evalString(src);
        } catch (LuaException e) {
            e.printStackTrace();
            Log.d("Vove :", "safeEvalLua  ----> " + e.getMessage());
        }
    }

    private void notifyOutput(int l, String o) {
        synchronized (printList) {
            for (LuaPrinter.OnPrint p : printList) {
                p.onPrint(l, o);
            }
        }
    }

    void evalString(String src, Object... args) throws LuaException {
        L.setTop(0);
        int ok = L.LloadString(src);
        if (ok == 0) {
            L.getGlobal("debug");
            L.getField(-1, "traceback");
            L.remove(-2);
            L.insert(-2);
            int l = args.length;
            for (Object arg : args) {
                L.pushObjectValue(arg);
            }
            ok = L.pcall(l, 0, -2 - l);
            if (ok == 0) {
                return;
            } else {
                String e = errorReason(ok) + ": " + L.toString(-1);
                Log.e("Vove :", "evalString  ----> " + e);
                notifyOutput(LuaManagerI.E, e);
                return;
            }
        }
        throw new LuaException(errorReason(ok) + ": " + L.toString(-1));
    }


    public void autoRun(String s, Object... args) throws LuaException {
        Log.d("Vove :", "autoRun  ----> " + s);
        if (Pattern.matches("^\\w+$", s)) {
            execFromAsset(s + ".lua", args);
        } else if (Pattern.matches("^[\\w._/]+$", s)) {
            execFromFile(s, args);
        } else {
            evalString(s, args);
        }
    }

    public void execFromFile(String filePath, Object... args) throws LuaException {
        int ok = 0;
        L.setTop(0);
        ok = L.LloadFile(filePath);

        if (ok == 0) {
            L.getGlobal("debug");
            L.getField(-1, "traceback");
            L.remove(-2);
            L.insert(-2);
            int l = args.length;
            for (int i = 0; i < l; i++) {
                L.pushObjectValue(args[i]);
            }
            ok = L.pcall(l, 0, -2 - l);
            if (ok == 0) {
                return;
            }
        }
        throw new LuaException(errorReason(ok) + ": " + L.toString(-1));
    }

    public void execFromAsset(String name, Object... args) throws LuaException {
        byte[] bytes;
        try {
            bytes = LuaUtil.readAsset(context, name);
        } catch (IOException e) {
            throw new LuaException(e.getMessage());
        }
        L.setTop(0);
        int ok = L.LloadBuffer(bytes, name);

        if (ok == 0) {
            L.getGlobal("debug");
            L.getField(-1, "traceback");
            L.remove(-2);
            L.insert(-2);
            int l = args.length;
            for (Object arg : args) {
                L.pushObjectValue(arg);
            }
            ok = L.pcall(l, 0, -2 - l);
            if (ok == 0) {
                return;
            }
        }
        throw new LuaException(errorReason(ok) + ": " + L.toString(-1));
    }

    @Override
    public void regGc(LuaGcable obj) {
        gcList.add(obj);
    }


    @Override
    public HashMap<String, String> getLibrarys() {
        return mLuaDexLoader.getLibrarys();
    }

    @Override
    public ArrayList<ClassLoader> getClassLoaders() {
        return mLuaDexLoader.getClassLoaders();
    }

    @Override
    public DexClassLoader loadDex(String path) throws LuaException {
        return mLuaDexLoader.loadDex(path);
    }

    @Override
    public void gc(LuaGcable obj) {
        obj.gc();
        gcList.remove(obj);
    }

    private void gcAll() {
        //清空线程..
        for (LuaGcable gcable : gcList) {
            if (gcable != null)
                gcable.gc();
        }
        gcList.clear();
    }

    @Override
    public void stop() {
        gcAll();
        L.gc(LUA_GCSTOP, 0);
        LuaStateFactory.removeLuaState(L.getPointer());
        L = null;
        init();
    }


    public void loadResources(String path) {
        mLuaDexLoader.loadResources(path);
    }

    /**
     * 注册打印
     */
    private final List<LuaPrinter.OnPrint> printList = new ArrayList<>();

    public void regPrint(LuaPrinter.OnPrint print) {
        printList.add(print);
    }

    @Override
    public Context getApp() {
        return context;
    }

    public void unRegPrint(LuaPrinter.OnPrint print) {
        printList.remove(print);
    }
}

File: androlua/src/main/java/cn/vove7/androlua/LuaHelper.kt
package cn.vove7.androlua

import android.content.Context
import cn.vove7.androlua.luabridge.LuaUtil
import cn.vove7.androlua.luabridge.LuaUtil.errorReason
import cn.vove7.androlua.luautils.LuaDexLoader
import cn.vove7.androlua.luautils.LuaGcable
import cn.vove7.androlua.luautils.LuaManagerI
import cn.vove7.androlua.luautils.LuaPrinter
import cn.vove7.common.ScriptEnginesBridges
import cn.vove7.common.MessageException
import cn.vove7.common.NotSupportException
import cn.vove7.common.app.GlobalApp
import cn.vove7.common.app.GlobalLog
import cn.vove7.common.executor.OnPrint
import cn.vove7.common.interfaces.ScriptEngine
import cn.vove7.vtp.log.Vog
import com.luajava.LuaException
import com.luajava.LuaState
import com.luajava.LuaState.LUA_GCSTOP
import com.luajava.LuaStateFactory
import dalvik.system.DexClassLoader
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentSkipListSet
import java.util.regex.Pattern
import kotlin.collections.HashSet

/**
 * LuaHelper
 *
 *
 * Created by Vove on 2018/7/30
 */
class LuaHelper : LuaManagerI, ScriptEngine {
    private val context: Context

    constructor(context: Context) {
        this.context = context
        bridgeManager = sScriptEnginesBridges
        initPath()
        init()
    }

    override var bridgeManager: ScriptEnginesBridges?


    constructor(context: Context, b: ScriptEnginesBridges) {
        this.context = context
        sScriptEnginesBridges = b
        bridgeManager = b
        initPath()
        init()
    }

    lateinit var L: LuaState
    override val luaState: LuaState
        get() = L

    private val gcList = ConcurrentSkipListSet<LuaGcable>()
    private var luaRequireSearchPath: String? = null
    private var mLuaDexLoader: LuaDexLoader? = null


    /**
     * 注册打印
     */
    companion object {
        private var libDir: String? = null
        private var jniLibsPath: String? = null
        private var luaDir: String? = null
        private val printList = HashSet<OnPrint>()
        private var sScriptEnginesBridges: ScriptEnginesBridges? = null

        fun regPrint(print: OnPrint) {
            synchronized(printList) {
                printList.add(print)
            }
        }

        fun unRegPrint(print: OnPrint) {
            synchronized(printList) {
                printList.remove(print)
            }
        }
    }

    private fun init() {
        synchronized(gcList) {
            gcList.clear()
        }
        mLuaDexLoader = LuaDexLoader(context)
        try {
            mLuaDexLoader!!.loadLibs()
        } catch (e: LuaException) {
            e.printStackTrace()
        }
        initLua()
    }

    private fun initPath() {
        luaDir = context.filesDir.absolutePath
        Vog.d("initLua luaDir ----> " + luaDir!!)
        libDir = context.getDir("lib", Context.MODE_PRIVATE).absolutePath
        jniLibsPath = context.applicationInfo.nativeLibraryDir + "/lib?.so" + ";" + libDir + "/lib?.so;"

        luaRequireSearchPath = "$luaDir/?.lua;$luaDir/lua/?.lua;$luaDir/?/init.lua;"
        luaRequireSearchPath += (jniLibsPath ?: ";")
    }

    override fun log(log: Any?) {
        GlobalLog.log(log.toString())
    }

    private fun initLua() {
        L = LuaStateFactory.newLuaState()
        L.openLibs()
        L.pushJavaObject(this)

        L.setGlobal("luaman")
        L.getGlobal("luaman")

        L.pushContext(context)
        //L.setGlobal("app");

//        L.getGlobal("luajava")

        //L.pushString(luaDir);
        //L.setField(-2, "luaextdir");
        L.pushString(luaDir)
        L.setField(-2, "luaDir")
        //L.pushString(luaDir);
        //L.setField(-2, "luapath");
        L.pop(1)

        L.getGlobal("package")
        L.pushString(luaRequireSearchPath)
        L.setField(-2, "path")
        L.pushString(jniLibsPath)
        L.setField(-2, "cpath")
        L.pop(1)

        LuaPrinter(L, object : OnPrint {
            override fun onPrint(l: Int, output: String?) {
                notifyOutput(l, output)
            }
        })

        //从assert加载api
        requireFromAsset(
                "lua_requires/import.lua",
                "lua_requires/bridges.lua",
                "lua_requires/utils.lua",
                "lua_requires/executor.lua",
                "lua_requires/view_op.lua",
                "lua_requires/global_op.lua",
                "lua_requires/storages.lua",
                "lua_requires/json.lua",
                "lua_requires/settings.lua"
        )
    }

    private fun requireFromAsset(vararg files: String) {
        files.forEach {
            execFromAsset(it, emptyArray())
        }
    }

    override fun handleMessage(l: Int, msg: String) {
        notifyOutput(l, msg)
    }


    private fun notifyOutput(l: Int, o: String?) {
        synchronized(printList) {
            for (p in printList) {
                p.onPrint(l, o)
            }
        }
    }

    private fun loadString(s: String): Int {
        L.top = 0
        return L.LloadString(s)
    }

    private fun loadFile(fileName: String): Int {
        L.top = 0
        return L.LloadFile(fileName)
    }

    @Throws(Exception::class)
    override fun evalString(script: String, args: Array<*>?) {
        val r = loadString(script)
        if (r == 0) {
            setArgs(args)
            exec(args?.size ?: 0)
        } else
            throw checkErr(r)
    }


    @Throws(Exception::class)
    override fun evalString(script: String, argMap: Map<String, *>?) {
        evalString(script, arrayOf(argMap))
    }

    override fun evalFile(file: String, argMap: Map<String, *>?) {
        evalFile(file, arrayOf(argMap))
    }

    private fun setArgs(args: Array<*>?) {
        L.getGlobal("debug")
        L.getField(-1, "traceback")
        L.remove(-2)
        L.insert(-2)

        args ?: return
        for (arg in args) {
            L.pushObjectValue(arg)
        }
    }

    private fun exec(argSize: Int) {
        val ok = L.pcall(argSize, 0, -2 - argSize)
        if (ok == 0) {
            return
        } else {//载入参数错误
            throw checkErr(ok)
        }
    }

    fun checkErr(r: Int): Throwable {
        val e = errorReason(r) + ": " + L.toString(-1)

        if (e.contains("java.lang.UnsupportedOperationException") ||
                e.contains("java.lang.InterruptedException")) {
            return InterruptedException(e)
        } else if (e.contains("cn.vove7.common.NotSupportException")) {
            return NotSupportException()
        }
        return MessageException(e)
    }

    override fun handleError(err: String) {
        GlobalLog.err(err)
        handleMessage(OnPrint.ERROR, err)
    }

    override fun handleError(e: Throwable) {
        GlobalLog.err(e)
        handleMessage(OnPrint.ERROR, e.message ?: "")
    }


    fun autoRun(s: String, args: Array<Any>) {
        Vog.d("autoRun  ----> $s")
        when {
            Pattern.matches("^\\w+$", s) -> execFromAsset("$s.lua", args)
            Pattern.matches("^[\\w._/]+$", s) -> evalFile(s, args)
            else -> evalString(s, args)
        }
    }

    override fun evalFile(file: String, args: Array<*>?) {
        var r = 1
        if (loadFile(file).also { r = it } == 0) {
            setArgs(args)
            exec(args?.size ?: 0)
        } else
            checkErr(r)
    }

    fun execFromAsset(name: String, args: Array<Any>? = null) {
        val bytes: ByteArray
        try {
            bytes = LuaUtil.readAsset(context, name)
        } catch (e: IOException) {
            e.printStackTrace()
            handleError(e)
            return
        }
        L.top = 0
        val ok = L.LloadBuffer(bytes, name)
        if (ok == 0) {
            setArgs(args)
            exec(args?.size ?: 0)
        } else checkErr(ok)
    }

    override fun regGc(obj: LuaGcable) {
        Vog.d("regGc $obj")
        synchronized(gcList) {
            gcList.add(obj)
        }
    }

    override val librarys: HashMap<String, String>
        get() = mLuaDexLoader!!.librarys


    override val classLoaders: ArrayList<ClassLoader>
        get() = mLuaDexLoader!!.classLoaders


    override fun loadDex(path: String): DexClassLoader? {
        var r: DexClassLoader? = null
        try {
            r = mLuaDexLoader!!.loadDex(path)
        } catch (e: Exception) {
            e.printStackTrace()
            handleError(e)
        }
        return r
    }

    override fun gc(obj: LuaGcable) {
        synchronized(gcList) {
            obj.gc()
            gcList.remove(obj)
        }
    }

    override fun removeGc(obj: LuaGcable): Boolean {
        synchronized(gcList) {
            return gcList.remove(obj).also {
                Vog.d("removeGc $obj $it")
            }
        }
    }

    private fun gcAll() {
        //清空线程..
        synchronized(gcList) {
            Vog.d("gcAll ${gcList.size}")
            for (gcable in gcList) {
                gcable.gc()
            }
            gcList.clear()
        }
    }

    override fun stop() {
        Vog.d("stop by external")
        release()
    }

    override fun release() {
        gcAll()
        L.gc(LUA_GCSTOP, 0)
        L.close()
    }

    fun loadResources(path: String) {
        mLuaDexLoader!!.loadResources(path)
    }

    override val app: Context
        get() = GlobalApp.APP

}
